docs/concepts/types.md

```diff
@@ -122,6 +122,49 @@ except ValidationError as exc:
     """
 ```
 
+#### Experimental pipeline API
+
+Pydantic v2.8.0 introduced an experimental pipeline API that allows composing parsing, constraints and transformations in a more type-safe manner than existing APIs. This API is subject to change or removal, we are looking for feedback and suggestions before making it a permanent part of Pydantic.
+
+```python
+from __future__ import annotations
+
+from datetime import datetime
+
+from typing_extensions import Annotated
+
+from pydantic import BaseModel
+from pydantic.transform_experimental import parse, parse_defer
+
+
+class User(BaseModel):
+    name: Annotated[str, parse(str).str.lower()]  # (1)!
+    age: Annotated[int, parse(int).gt(0)]  # (2)!
+    username: Annotated[str, parse(str).str.pattern(r'[a-z]+')]  # (3)!
+    password: Annotated[
+        str,
+        parse(str).transform(str.lower).predicate(lambda x: x != 'password')]  # (4)!
+    favorite_number: Annotated[  # (5)!
+        int,
+        (parse(int) | parse(str).str.strip().parse(int)).gt(0),
+    ]
+    friends: Annotated[list[User], parse().len(0, 100)]  # (6)!
+    family: Annotated[  # (7)!
+        list[User],
+        parse_defer(lambda: list[User]).transform(lambda x: x[1:]),
+    ]
+    bio: Annotated[datetime, parse(int).transform(lambda x: x / 1_000_000).parse()]  # (8)!
+```
+
+1. Lowercase a string.
+2. Constrain an integer to be greater than zero.
+3. Constrain a string to match a regex pattern.
+4. You can also use the lower level transform, constrain and predicate methods.
+5. Use the `|` or `&` operators to combine steps (like a logical OR or AND).
+6. Calling `parse()` with no arguments implies `parse(<field type>)`. Use `parse(Any)` to accept any type.
+7. For recursive types you can use `parse_defer` to reference the type itself before it's defined.
+8. You can call `parse()` before or after other steps to do pre or post processing.
```

Comment:
Can you explain what this means? I assume that this is interactions with `BeforeValidator`. How do we implement wrap validators in this model?

Reply:
Oh I see. `.parse()` like that was totally not obvious to me, is it a significant ergonomic win to allow the argument to be inferred. Otherwise I might argue to make it required for now to simplify.

Or write an explicit section which is like

```
BeforeValidator -> parse(str).transform(bv).parse()
AfterValidator -> parse().transform(av)
WrapValidator -> parse(str).transform(bv).parse().transform(av)
```

... @sydney-runkle and I had an interesting use case for a wrap validator the other day which split IANA timezone off a timestamp so that we could do the timestamp in Rust and then reconstruct a tz-aware datetime. It would be interesting to see how to implement that in this API, as it's not obvious to me.

Reply:
Could you elaborate on that example and we can try and see how it would work out?

Reply:
> is it a significant ergonomic win to allow the argument to be inferred
I'd say so, consider if the type is `list[MyGeneric[dict[str, set[bool]]]]` or something...

Reply:
> had an interesting use case for a wrap validator the other day which split IANA timezone off a timestamp so that we could do the timestamp in Rust and then reconstruct a tz-aware datetime

We could also add an explicit wrap validator: `parse(str).transform(lambda x, h: h(x.strip())).parse(int)`?